src/Stratis.Bitcoin.Features.Consensus/Rules/ProvenHeaderRules/ProvenHeaderCoinstakeRule.cs

```diff
@@ -0,0 +1,281 @@
+
+            Guard.NotNull(this.PosParent.StakeValidator, nameof(this.PosParent.StakeValidator));
+            Guard.NotNull(this.PosParent.UtxoSet, nameof(this.PosParent.UtxoSet));
+
+            this.stakeValidator = this.PosParent.StakeValidator;
+        }
+
+        /// <inheritdoc/>
+        public override void Run(RuleContext context)
+        {
+            Guard.NotNull(context.ValidationContext.ChainedHeaderToValidate, nameof(context.ValidationContext.ChainedHeaderToValidate));
+
+            ChainedHeader chainedHeader = context.ValidationContext.ChainedHeaderToValidate;
+            int height = chainedHeader.Height;
+
+            if (context.SkipValidation || !this.IsProvenHeaderActivated(height))
+                return;
+
+            var header = (ProvenBlockHeader)chainedHeader.Header;
+
+            this.CheckCoinstakeIsNotNull(header);
+
+            this.CheckIfCoinstakeIsTrue(header);
+
+            this.CheckHeaderAndCoinstakeTimes(header);
+
+            FetchCoinsResponse coins = this.GetAndValidateCoins(header);
+            UnspentOutputs prevUtxo = this.GetAndValidatePreviousUtxo(coins);
+
+            this.CheckCoinstakeAgeRequirement(chainedHeader, prevUtxo);
+
+            this.CheckSignature(header, prevUtxo);
+
+            this.CheckStakeKernelHash((PosRuleContext)context, prevUtxo, header, chainedHeader);
+
+            this.CheckCoinstakeMerkleProof(header);
+
+            this.CheckHeaderSignatureWithCoinstakeKernel(header, prevUtxo);
+        }
+
+        /// <summary>
+        /// Checks the coinstake to make sure it is not null.
+        /// </summary>
+        /// <param name="header">The proven blockheader.</param>
+        /// <exception cref="ConsensusException">
+        /// Throws exception with error <see cref="ConsensusErrors.EmptyCoinstake" /> if check fails.
+        /// </exception>
+        private void CheckCoinstakeIsNotNull(ProvenBlockHeader header)
+        {
+            if (header.Coinstake != null)
+                return;
+
+            this.Logger.LogTrace("(-)[COINSTAKE_IS_NULL]");
+            ConsensusErrors.EmptyCoinstake.Throw();
+        }
+
+        /// <summary>
+        /// Fetches and validates coins from coins view.
+        /// </summary>
+        /// <param name="header">The header.</param>
+        /// <exception cref="ConsensusException">
+        /// Throws exception with error <see cref="ConsensusErrors.ReadTxPrevFailed" /> if check fails.
+        /// </exception>
+        private FetchCoinsResponse GetAndValidateCoins(ProvenBlockHeader header)
+        {
+            // First try finding the previous transaction in database.
+            TxIn txIn = header.Coinstake.Inputs[0];
+            FetchCoinsResponse coins = this.PosParent.UtxoSet.FetchCoinsAsync(new[] { txIn.PrevOut.Hash }).GetAwaiter().GetResult();
+            if ((coins == null) || (coins.UnspentOutputs.Length != 1))
+                ConsensusErrors.ReadTxPrevFailed.Throw();
```

Comment:
As I am given to understand this will mark the header before it as consumed on the tree and we will not continue validation until CT advances and the UTXO is present.